packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+    // inflate the children in reverse tree order.
+
+    if (currentLogicalIndex > 0 && currentLogicalOffset > startLogicalOffset) {
+      final List<Element> newChildren = <Element>[];
+
+      while (currentLogicalIndex > 0 && currentLogicalOffset > startLogicalOffset) {
+        currentLogicalIndex -= 1;
+        Widget newWidget = new RepaintBoundary.wrap(builder(this, currentLogicalIndex), currentLogicalIndex);
+        assert(newWidget != null);
+        newChildren.add(inflateWidget(newWidget, null));
+        RenderBox child = block.firstChild;
+        assert(child == newChildren.last.renderObject);
+        child.layout(innerConstraints, parentUsesSize: true);
+        currentLogicalOffset -= _getMainAxisExtent(child.size);
+      }
+
+      final int numberOfNewChildren = newChildren.length;
+      _children.insertAll(0, newChildren.reversed);
+      _firstChildLogicalIndex = currentLogicalIndex;
+      _firstChildLogicalOffset = currentLogicalOffset;
+      firstLogicalIndexNeedingLayout = currentLogicalIndex + numberOfNewChildren;
+    } else if (currentLogicalOffset < startLogicalOffset) {
+      // If we didn't need to inflate more children before the viewport, we
+      // might need to deactivate children that have left the viewport from the
+      // top. We repeatedly check whether the first child overlaps the viewport
+      // and deactivate it if it's outside the viewport.
+      int currentPhysicalIndex = 0;
+      while (block.firstChild != null) {
+        RenderBox child = block.firstChild;
+        child.layout(innerConstraints, parentUsesSize: true);
+        firstLogicalIndexNeedingLayout += 1;
+        double childExtent = _getMainAxisExtent(child.size);
+        if (currentLogicalOffset + childExtent >= startLogicalOffset)
+          break;
+        deactivateChild(_children[currentPhysicalIndex]);
+        _children[currentPhysicalIndex] = null;
+        currentPhysicalIndex += 1;
+        currentLogicalIndex += 1;
+        currentLogicalOffset += childExtent;
+      }
+
+      if (currentPhysicalIndex > 0) {
+        _children.removeRange(0, currentPhysicalIndex);
+        _firstChildLogicalIndex = currentLogicalIndex;
+        _firstChildLogicalOffset = currentLogicalOffset;
+      }
+    }
+
+    // We've now established the invariant that the first physical child in the
+    // block is the first child that ought to be visible in the viewport. Now we
+    // need to walk forward until we've filled up the viewport. We might have
+    // already called layout for some of the children we encounter in this phase
+    // of the algorithm, we we'll need to be careful not to call layout on them again.
+
+    if (currentLogicalOffset >= startLogicalOffset) {
+      // The first element is visible. We need to update our reckoning of where
+      // the min scroll offset is.
+      _minScrollOffset = currentLogicalOffset;
+      _startOffsetLowerLimit = double.NEGATIVE_INFINITY;
```

Comment:
this code confuses me.

The first element is visible if currentLogicalOffset is greater than startLogicalOffset. startLogicalOffset is ideally zero. This means currentLogicalOffset being positive indicates we are underscrolled. This means we can make it more and more positive and we won't need to rebuild. So then why are we setting the _lower_ limit rather than the _upper_ limit?